Manage local models through application

jelmervdl opened the pull request:
Solving #45.

Idea is to:
- [x] Add a table to the settings dialog to see & delete local models
  - [x] only the ones in .config/translateLocally can be deleted. Do not attempt to delete models found in `pwd`.
  - [x] Make delete option prominent so the intention of that table becomes immediately clear  
_I'm thinking adding a column with a delete button per row, but might end up with a button at the bottom of the table that is only enabled if deletable modes are selected. Like how the context menu works. Because of space._
- [x] Add an import button to load (copy) local models into the local "library" (the .config/translateLocally folder)
  - [x] validate before importing (check whether there is a model_json file?)  
_partially done, model is not imported but extracted files stay_

This won't affect:
- Reading local models from the start-up directory. That should keep working. Those models should not be imported into the local library.

XapaJIaMnu commented:
I compiled and tried.
1) We should do something about auto-resizing the widget when switching to the languages tab, as the user needs to resize things to see the full name.
2) We should give the user the information why some models are undeleteable (eg, they were distributed with the exe, and the user can just remove them from the current directory)

In terms of validation, we this is already done by this function: 
https://github.com/XapaJIaMnu/translateLocally/blob/db88f0d571055ae7750a72f35413575e2dad6f6f/src/ModelManager.cpp#L144
Difference being, it takes a parent folder. Maybe some of it can be abstracted to work on a specific model file. (OR you know use this function: https://github.com/XapaJIaMnu/translateLocally/blob/db88f0d571055ae7750a72f35413575e2dad6f6f/src/ModelManager.cpp#L37 )

As a side note, we probably should explain in more details the change dir shenanigans that go on in here: https://github.com/XapaJIaMnu/translateLocally/blob/db88f0d571055ae7750a72f35413575e2dad6f6f/src/ModelManager.cpp#L189

XapaJIaMnu commented on src/TranslatorSettingsDialog.ui:
```diff
@@ -6,63 +6,155 @@
     <y>0</y>
-    <width>300</width>
-    <height>244</height>
+    <width>362</width>
```
Every time one of us works on QtCreator, we have phantom resizes here and there. Not sure how we can completely avoid that.

jelmervdl replied:
Probably by being a bit more strict in which lines we commit. This time it was intentional though, 300px was a bit too small for the dialog to also contain the tab pane and the table. 362 is a bit random…

XapaJIaMnu replied:
Not a criticism, just an observation. I've noticed this with my commits as well.

jelmervdl commented:
A problem with the current approach to models is that we really trust the .tar.gz files.

We assume
- that the name of the .tar.gz file is also the name of the folder it contains
- that it contains a single folder with all the model files
- that no two different models shall have the same folder name
- that it doesn't contain any symlinks or paths with ".." in it (I've seen that libarchive has some flags to guarantee this, but we don't explicitly set them)

Additionally, right now if you upgrade a model it does so by extracting over the current model. This could leave files from the old model that haven't been overwritten lying around. It also corrupts the old model when the updated model fails to extract completely.

I think we could solve all of the above if:
- We always extract models to a unique folder (say like UUID based folder names)
- If all entries in the .tar.gz indeed have the same prefix folder, we substitute that for the unique folder. Otherwise we extract all files as is to the unique folder (so models can still have subdirectories if they need to)
- If extraction fails half-way through, or the model_info.json file is not found, we can just delete the unique folder
- If it does extract & loads correctly, and it is the same model (same long name?) but a newer version of an already installed model, we can delete the old model completely by just removing that unique folder. It won't affect the updated model.

jelmervdl commented:
Additionally we might want to add checksum to the model index at some point? Then we can just point to models on S3 or 3rd party sites and validate the .tar.gz files before extracting them. Won't solve the above issues for locally imported models (e.g. by using the "Import models…" button I added.

XapaJIaMnu commented:
I'm all for md5 checksums being distributed alongside the tar.gz possibly inside the models.json which we fetch from the server. Then we can download and compute the md5sum.

Can we not do opaque folder names? We could add a timestamp to them, I guess. I see little benefit of it though, we could just delete the previous folder before extracting?

kpu commented:
md5 is broken.  Let's do something more secure.

jelmervdl commented:
Any of these would be fine with me: https://doc.qt.io/qt-5/qcryptographichash.html#Algorithm-enum

jelmervdl commented:
> Can we not do opaque folder names?

I don't trust the archives to always have a folder with a unique name in them, or a consistent name across upgrades. Like, that would be okay if it was just a single source for models. But I just added a button to import your own .tar.gz files, opening it up for all kinds of shenanigans.


> we could just delete the previous folder before extracting?

I'm thinking of the case extraction might fail, or it might be an incomplete archive. Then we also just deleted the old working version.

> We could add a timestamp to 
Them

That would be sufficient to get around the above failure case

XapaJIaMnu commented:
We are the single source models for now, but yeah, we might not be in the future. I don't see users being too keen to prepare their own models, but hopefully that will change. And when they do, they would expect models with the same folder names to override the previous model they tried to import. I guess confirmation for "Folder X, exists, replace?" would work here.
We also have to teach people how to prepare their `model_info.json`, I guess.

I'm not sure adding an extra timestamp to the folder name is absolutely necessary. It is defnitely not harmful.

XapaJIaMnu commented:
Delete models was somewhat buggy for me. I had 5 models and I deleted them one by one and at first, multiple models disappeared at once, then later on I kept having 1 model left, but as I deleted it, another one appeared. Could it because I had the .tar.gz files inside the directory, as per the behaviour of the master branch?

The initial model manager window is a bit tiny, especially if one has many models installed.

XapaJIaMnu commented:
Same run, when I try to download a new model after deleting everything:
![Screenshot_20210624_183123](https://user-images.githubusercontent.com/2027221/123307570-8d372700-d51a-11eb-8a99-037bfcf194e2.png)

jelmervdl commented:
Oooh interesting! Why would it not be able to move the folder? Hmm… I'll experiment a bit more.

I'll also look into the weird deleting/popping back models. Probably a signalling issue in ModelManager. Making the QAbstractTableModel emit the right signals is a bit tricky.

XapaJIaMnu commented:
Same issue, if I try to import a model. I assume it works for you on mac?

jelmervdl commented:
Yeah, worked fine for me. Not sure why it doesn't work for linux. Does the `~/.config/translateLocally` path exist when you encounter this error?

XapaJIaMnu commented:
Yes, it does exist. It is created automatically on the first run if it doesn't exist.

jelmervdl commented:
I have no clue why it would not just work. QDir::rename() and QTemporaryDir both don't do anything magical. I'll try it in my Ubuntu VM.

jelmervdl commented:
I haven't been able to replicate the rename() failure. What OS are trying this on (looks like something with KDE)

Do you have any special setup around your temp and home directory? Like, could there be any reason why a directory in /tmp could not be moved to ~/.config/translateLocally? Different filesystem maybe?

Lastly, could you try the commented out line in ModelManger.cpp that creates a temp directory in the translateLocally directory directly, and see if that solves it? I.e. change ModelManager.cpp:91 to:
```cpp
// Initially extract to to a temporary directory. Will delete its contents
// when it goes out of scope
QTemporaryDir tempDir(configDir_.filePath(".extracting-XXXXXXX"));
// QTemporaryDir tempDir;
```

XapaJIaMnu commented:
I'll take a look over the weekend. I suspect it might be an issue with the working directory not being clean due to it being created by a previous version of translateLocally. 

I use arch Linux

XapaJIaMnu commented:
I think i found the issue. Across SO:
> ...
> Do remember that even with giving a full path QDir::rename will fail if source and destination path's are not on the same partition or for a few more reasons as mentioned in the documentation.
> ...
My tmp is not on the same partition.

XapaJIaMnu commented:
This will potentially be a problem on a number of systems that use different drive configurations. With some brief googling, it seems everyone recommends the use of `QDir::rename` which will fail in my case. Not sure what's the next best thing to do. Maintain our own `temp` in the configuration dir? Use C++17 filesystem?

jelmervdl commented:
I think the safest work-around is just creating a temp directory inside the config dir. This also saves us from creating temp files and then ending up trying to move them to a filesystem that's full or has a quota on it that causes the move to fail.

I've explicitly excluded any of the temp directories from being scanned. Not that they should still exist at that point, but who knows what weird crashes or bugs may cause them to survive their pruning. I thought about just making them hidden directories and excluding them that way from the scan in `scanForModels()`, but I'm not sure that works on Windows.

XapaJIaMnu commented:
Yes, this finally works for me.

How about deleting any `extracting-XXXX` directories when the program is run? That should do a proper cleanup in case we had a crash during model extraction on a previous run.

XapaJIaMnu commented:
Or is this done automatically by the temporary dir function?

jelmervdl commented:
Yes. the QTemporaryDir will delete the directory it made (recursively) when it's deallocated. It's only when the application segfaults (e.g. somewhere in libarchive) that the `extracting-XXXX` folder would stay.

I'm not a fan of the idea of deleting these folders on start-up. I think Linux and Windows there's nothing preventing you from starting a second instance of the program. If you would do that exactly while another instance is extracting, it would break that process.

XapaJIaMnu commented:
Fair point, merging.
